# Data validation & sanity check

Setup & Imports

In [1]:
# ======================
# Setup
# ======================
import sys
import os

sys.path.append(os.path.abspath(".."))

import pandas as pd
import numpy as np

from src.data_loader import load_data
from src.config import set_seed

set_seed(42)

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:.4f}".format)

In [ ]:
# Load Data

train, test, econ = load_data()

print("Train shape:", train.shape)
print("Test shape:", test.shape)
print("Economic shape:", econ.shape)

Train shape: (68654, 16)
Test shape: (18594, 15)
Economic shape: (27, 25)


3. Dataset Overview

In [ ]:
def dataset_overview(df, name):
    print(f"\n{'='*50}")
    print(f"{name} OVERVIEW")
    print(f"{'='*50}")
    
    print("\nShape:", df.shape)
    print("\nColumns:")
    print(df.columns.tolist())
    
    print("\nHead:")
    display(df.head())


dataset_overview(train, "TRAIN")
dataset_overview(test, "TEST")
dataset_overview(econ, "ECONOMIC DATA")


TRAIN OVERVIEW

Shape: (68654, 16)

Columns:
['ID', 'customer_id', 'country_id', 'tbl_loan_id', 'lender_id', 'loan_type', 'Total_Amount', 'Total_Amount_to_Repay', 'disbursement_date', 'due_date', 'duration', 'New_versus_Repeat', 'Amount_Funded_By_Lender', 'Lender_portion_Funded', 'Lender_portion_to_be_repaid', 'target']

Head:


,ID,customer_id,country_id,tbl_loan_id,lender_id,loan_type,Total_Amount,Total_Amount_to_Repay,disbursement_date,due_date,duration,New_versus_Repeat,Amount_Funded_By_Lender,Lender_portion_Funded,Lender_portion_to_be_repaid,target
0,ID_257244237466267278,257244,Kenya,237466,267278,Type_1,16477.0000,16642.0000,2022-08-15,2022-08-22,7,Repeat Loan,2436.0000,0.1478,2460.0000,0
1,ID_249619214444267278,249619,Kenya,214444,267278,Type_1,5599.0000,5599.0000,2022-07-13,2022-07-20,7,Repeat Loan,2799.5000,0.5000,2800.0000,0
2,ID_251457257838267278,251457,Kenya,257838,267278,Type_1,4679.0000,4746.0000,2022-09-12,2022-09-19,7,Repeat Loan,7.8000,0.0017,8.0000,0
3,ID_266224270337267278,266224,Kenya,270337,267278,Type_1,6123.0000,6123.0000,2022-09-29,2022-10-06,7,Repeat Loan,318.8000,0.0521,319.0000,0
4,ID_250369246185267278,250369,Kenya,246185,267278,Type_1,2953.0000,2974.0000,2022-08-27,2022-09-03,7,Repeat Loan,15.0000,0.0051,15.0000,0



TEST OVERVIEW

Shape: (18594, 15)

Columns:
['ID', 'customer_id', 'country_id', 'tbl_loan_id', 'lender_id', 'loan_type', 'Total_Amount', 'Total_Amount_to_Repay', 'disbursement_date', 'due_date', 'duration', 'New_versus_Repeat', 'Amount_Funded_By_Lender', 'Lender_portion_Funded', 'Lender_portion_to_be_repaid']

Head:


,ID,customer_id,country_id,tbl_loan_id,lender_id,loan_type,Total_Amount,Total_Amount_to_Repay,disbursement_date,due_date,duration,New_versus_Repeat,Amount_Funded_By_Lender,Lender_portion_Funded,Lender_portion_to_be_repaid
0,ID_263861241254267278,263861,Kenya,241254,267278,Type_1,6990.0000,7032.0000,8/20/2022,8/27/2022,7,Repeat Loan,948.1900,0.1356,954.0000
1,ID_253607258240267278,253607,Kenya,258240,267278,Type_1,5319.0000,5446.0000,9/13/2022,9/20/2022,7,Repeat Loan,0.0000,0.0000,0.0000
2,ID_248459265311267278,248459,Kenya,265311,267278,Type_1,3499.0000,3499.0000,9/22/2022,9/29/2022,7,Repeat Loan,5.2600,0.0015,5.0000
3,ID_258351279506267278,258351,Kenya,279506,267278,Type_1,6808.0000,7017.0000,10/11/2022,10/18/2022,7,Repeat Loan,938.7800,0.1379,968.0000
4,ID_262673283885267278,262673,Kenya,283885,267278,Type_1,13479.0000,13807.0000,10/18/2022,10/25/2022,7,Repeat Loan,4043.7000,0.3000,4142.0000



ECONOMIC DATA OVERVIEW

Shape: (27, 25)

Columns:
['Country', 'Indicator', 'YR2001', 'YR2002', 'YR2003', 'YR2004', 'YR2005', 'YR2006', 'YR2007', 'YR2008', 'YR2009', 'YR2010', 'YR2011', 'YR2012', 'YR2013', 'YR2014', 'YR2015', 'YR2016', 'YR2017', 'YR2018', 'YR2019', 'YR2020', 'YR2021', 'YR2022', 'YR2023']

Head:


,Country,Indicator,YR2001,YR2002,YR2003,YR2004,YR2005,YR2006,YR2007,YR2008,YR2009,YR2010,YR2011,YR2012,YR2013,YR2014,YR2015,YR2016,YR2017,YR2018,YR2019,YR2020,YR2021,YR2022,YR2023
0,Ghana,"Inflation, consumer prices (annual %)",41.5095,9.3609,29.7730,18.0427,15.4390,11.6792,10.7343,16.4946,19.2469,10.7334,8.7285,11.1863,11.6662,15.4896,17.1500,17.4546,12.3719,7.8088,7.1436,9.8873,9.9711,31.2559,38.1070
1,Cote d'Ivoire,"Inflation, consumer prices (annual %)",4.3615,3.0773,3.2968,1.4580,3.8858,2.4672,1.8920,6.3085,1.0195,1.2265,4.9124,1.3045,2.5812,0.4487,1.2515,0.7232,0.6859,0.3594,-1.1069,2.4250,4.0920,5.2762,4.3871
2,Kenya,"Inflation, consumer prices (annual %)",5.7386,1.9613,9.8157,11.6240,10.3128,14.4537,9.7589,26.2398,9.2341,3.9614,14.0225,9.3778,5.7175,6.8782,6.5822,6.2972,8.0056,4.6898,5.2396,5.4052,6.1079,7.6599,7.6714
3,Ghana,"Official exchange rate (LCU per US$, period av...",0.7163,0.7924,0.8668,0.8995,0.9052,0.9151,0.9326,1.0523,1.4050,1.4300,1.5206,1.8249,1.9813,2.8966,3.7146,3.9098,4.3505,4.5853,5.2174,5.5957,5.8057,8.2724,11.0204
4,Cote d'Ivoire,"Official exchange rate (LCU per US$, period av...",732.3977,693.7132,579.8974,527.3380,527.2584,522.4256,478.6337,446.0000,470.2934,494.7943,471.2486,510.5563,493.8996,493.7573,591.2117,592.6056,580.6567,555.4465,585.9110,575.5860,554.5307,623.7597,606.5698


4. Schema and Data Types Validation

In [ ]:
def check_dtypes(df, name):
    print(f"\n{name} DATA TYPES")
    print(df.dtypes)

check_dtypes(train, "TRAIN")
check_dtypes(test, "TEST")


TRAIN DATA TYPES
ID                                 str
customer_id                      int64
country_id                         str
tbl_loan_id                      int64
lender_id                        int64
loan_type                          str
Total_Amount                   float64
Total_Amount_to_Repay          float64
disbursement_date                  str
due_date                           str
duration                         int64
New_versus_Repeat                  str
Amount_Funded_By_Lender        float64
Lender_portion_Funded          float64
Lender_portion_to_be_repaid    float64
target                           int64
dtype: object

TEST DATA TYPES
ID                                 str
customer_id                      int64
country_id                         str
tbl_loan_id                      int64
lender_id                        int64
loan_type                          str
Total_Amount                   float64
Total_Amount_to_Repay          float64
disbursement_da

5. Missing Values Analysis

In [5]:
def missing_values(df, name):
    print(f"\n{name} MISSING VALUES")
    
    missing = df.isnull().sum()
    missing = missing[missing > 0].sort_values(ascending=False)
    
    if missing.empty:
        print("No missing values ✅")
    else:
        display(pd.DataFrame({
            "missing_count": missing,
            "missing_pct": (missing / len(df)) * 100
        }))

missing_values(train, "TRAIN")
missing_values(test, "TEST")
missing_values(econ, "ECONOMIC")


TRAIN MISSING VALUES
No missing values ✅

TEST MISSING VALUES
No missing values ✅

ECONOMIC MISSING VALUES


,missing_count,missing_pct
YR2023,13,48.1481
YR2022,13,48.1481
YR2018,10,37.0370
YR2019,10,37.0370
YR2020,10,37.0370
YR2021,10,37.0370
YR2001,7,25.9259
YR2003,7,25.9259
YR2002,7,25.9259
YR2004,7,25.9259


6. Duplicate & Key Integrity Checks

In [6]:
def duplicate_checks(df, name):
    print(f"\n{name} DUPLICATE CHECKS")
    
    print("Full duplicates:", df.duplicated().sum())
    
    if "ID" in df.columns:
        print("Duplicate IDs:", df["ID"].duplicated().sum())

duplicate_checks(train, "TRAIN")
duplicate_checks(test, "TEST")


TRAIN DUPLICATE CHECKS
Full duplicates: 0
Duplicate IDs: 0

TEST DUPLICATE CHECKS
Full duplicates: 0
Duplicate IDs: 0


7. ID Relationship Analysis

In [7]:
print("\n🔍 ID RELATIONSHIP ANALYSIS")

print("\nUnique counts:")
print("Customers:", train["customer_id"].nunique())
print("Loans:", train["tbl_loan_id"].nunique())
print("Lenders:", train["lender_id"].nunique())

print("\nRows per loan (should show multi-lender structure):")
loan_counts = train.groupby("tbl_loan_id").size()
display(loan_counts.describe())

print("\nRows per customer:")
customer_counts = train.groupby("customer_id").size()
display(customer_counts.describe())


🔍 ID RELATIONSHIP ANALYSIS

Unique counts:
Customers: 6517
Loans: 66569
Lenders: 4

Rows per loan (should show multi-lender structure):


count   66569.0000
mean        1.0313
std         0.1742
min         1.0000
25%         1.0000
50%         1.0000
75%         1.0000
max         2.0000
dtype: float64


Rows per customer:


count   6517.0000
mean      10.5346
std       12.9728
min        1.0000
25%        2.0000
50%        6.0000
75%       14.0000
max      206.0000
dtype: float64

8. Target Variable Analysis

In [8]:
print("\n🎯 TARGET DISTRIBUTION")

target_counts = train["target"].value_counts()
print(target_counts)

print("\nPercentage:")
print(train["target"].value_counts(normalize=True) * 100)


🎯 TARGET DISTRIBUTION
target
0    67396
1     1258
Name: count, dtype: int64

Percentage:
target
0   98.1676
1    1.8324
Name: proportion, dtype: float64


9. Date Consistency Checks

In [9]:
date_cols = ["disbursement_date", "due_date"]

for col in date_cols:
    train[col] = pd.to_datetime(train[col], errors="coerce")
    test[col] = pd.to_datetime(test[col], errors="coerce")

print("\nDate conversion complete")

# Check negative durations
print("\nChecking date consistency...")
invalid_dates = (train["due_date"] < train["disbursement_date"]).sum()
print("Invalid date rows:", invalid_dates)


Date conversion complete

Checking date consistency...
Invalid date rows: 0


10. Numeric Sanity Checks

In [10]:
numeric_cols = train.select_dtypes(include=np.number).columns

print("\n🔢 NUMERIC SUMMARY")
display(train[numeric_cols].describe())

# Check negative values
for col in numeric_cols:
    if (train[col] < 0).sum() > 0:
        print(f"⚠️ Negative values in {col}")


🔢 NUMERIC SUMMARY


,customer_id,tbl_loan_id,lender_id,Total_Amount,Total_Amount_to_Repay,duration,Amount_Funded_By_Lender,Lender_portion_Funded,Lender_portion_to_be_repaid,target
count,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000,68654.0000
mean,254392.7493,262982.5806,266432.6526,14592.7048,15399.3403,8.5036,2501.3093,0.2191,2603.9781,0.0183
std,26623.5269,39328.0232,3567.1361,133668.3710,161157.2487,13.4266,11486.0661,0.1297,12916.5636,0.1341
min,145.0000,101323.0000,245684.0000,5.0000,0.0000,1.0000,0.0000,0.0000,0.0000,0.0000
25%,248963.0000,233890.5000,267278.0000,2289.0000,2321.0000,7.0000,239.0000,0.1213,242.0000,0.0000
50%,255410.0000,260246.5000,267278.0000,5248.0000,5299.0000,7.0000,909.0000,0.3000,927.0000,0.0000
75%,262332.0000,286898.7500,267278.0000,11388.0000,11596.0000,7.0000,2258.4000,0.3000,2303.0000,0.0000
max,312737.0000,375320.0000,267278.0000,20000000.0000,24152842.0000,1096.0000,1600000.0000,1.1681,1821338.0000,1.0000


11. Categorical Validation

In [11]:
categorical_cols = train.select_dtypes(include="object").columns

for col in categorical_cols:
    print(f"\nColumn: {col}")
    print(train[col].value_counts(dropna=False).head(10))


Column: ID
ID
ID_257244237466267278    1
ID_249619214444267278    1
ID_251457257838267278    1
ID_266224270337267278    1
ID_250369246185267278    1
ID_262778277126267278    1
ID_259611263472267278    1
ID_261867288216267278    1
ID_259525247727267278    1
ID_251756234445267278    1
Name: count, dtype: int64

Column: country_id
country_id
Kenya    68654
Name: count, dtype: int64

Column: loan_type
loan_type
Type_1     61823
Type_7      2749
Type_5      1514
Type_4      1203
Type_9       475
Type_6       360
Type_2       360
Type_10       45
Type_15       36
Type_14       20
Name: count, dtype: int64

Column: New_versus_Repeat
New_versus_Repeat
Repeat Loan    68112
New Loan         542
Name: count, dtype: int64


C:\Users\otsyu\AppData\Local\Temp\ipykernel_34084\1851763171.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = train.select_dtypes(include="object").columns


11. Categorical Validation

In [12]:
categorical_cols = train.select_dtypes(include="object").columns

for col in categorical_cols:
    print(f"\nColumn: {col}")
    print(train[col].value_counts(dropna=False).head(10))


Column: ID
ID
ID_257244237466267278    1
ID_249619214444267278    1
ID_251457257838267278    1
ID_266224270337267278    1
ID_250369246185267278    1
ID_262778277126267278    1
ID_259611263472267278    1
ID_261867288216267278    1
ID_259525247727267278    1
ID_251756234445267278    1
Name: count, dtype: int64

Column: country_id
country_id
Kenya    68654
Name: count, dtype: int64

Column: loan_type
loan_type
Type_1     61823
Type_7      2749
Type_5      1514
Type_4      1203
Type_9       475
Type_6       360
Type_2       360
Type_10       45
Type_15       36
Type_14       20
Name: count, dtype: int64

Column: New_versus_Repeat
New_versus_Repeat
Repeat Loan    68112
New Loan         542
Name: count, dtype: int64


C:\Users\otsyu\AppData\Local\Temp\ipykernel_34084\1851763171.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = train.select_dtypes(include="object").columns


12. Cross-Dataset Consistency (Train vs Test)

In [13]:
print("\n🔄 TRAIN vs TEST COLUMN MATCH")

train_cols = set(train.columns)
test_cols = set(test.columns)

print("Only in train:", train_cols - test_cols)
print("Only in test:", test_cols - train_cols)


🔄 TRAIN vs TEST COLUMN MATCH
Only in train: {'target'}
Only in test: set()


13. Economic Data Validation

In [14]:
print("\n🌍 ECONOMIC DATA CHECK")

print("Countries:", econ["Country"].unique())
print("Indicators:", econ["Indicator"].nunique())

print("\nSample:")
display(econ.head())


🌍 ECONOMIC DATA CHECK
Countries: <StringArray>
['Ghana', 'Cote d'Ivoire', 'Kenya']
Length: 3, dtype: str
Indicators: 9

Sample:


,Country,Indicator,YR2001,YR2002,YR2003,YR2004,YR2005,YR2006,YR2007,YR2008,YR2009,YR2010,YR2011,YR2012,YR2013,YR2014,YR2015,YR2016,YR2017,YR2018,YR2019,YR2020,YR2021,YR2022,YR2023
0,Ghana,"Inflation, consumer prices (annual %)",41.5095,9.3609,29.7730,18.0427,15.4390,11.6792,10.7343,16.4946,19.2469,10.7334,8.7285,11.1863,11.6662,15.4896,17.1500,17.4546,12.3719,7.8088,7.1436,9.8873,9.9711,31.2559,38.1070
1,Cote d'Ivoire,"Inflation, consumer prices (annual %)",4.3615,3.0773,3.2968,1.4580,3.8858,2.4672,1.8920,6.3085,1.0195,1.2265,4.9124,1.3045,2.5812,0.4487,1.2515,0.7232,0.6859,0.3594,-1.1069,2.4250,4.0920,5.2762,4.3871
2,Kenya,"Inflation, consumer prices (annual %)",5.7386,1.9613,9.8157,11.6240,10.3128,14.4537,9.7589,26.2398,9.2341,3.9614,14.0225,9.3778,5.7175,6.8782,6.5822,6.2972,8.0056,4.6898,5.2396,5.4052,6.1079,7.6599,7.6714
3,Ghana,"Official exchange rate (LCU per US$, period av...",0.7163,0.7924,0.8668,0.8995,0.9052,0.9151,0.9326,1.0523,1.4050,1.4300,1.5206,1.8249,1.9813,2.8966,3.7146,3.9098,4.3505,4.5853,5.2174,5.5957,5.8057,8.2724,11.0204
4,Cote d'Ivoire,"Official exchange rate (LCU per US$, period av...",732.3977,693.7132,579.8974,527.3380,527.2584,522.4256,478.6337,446.0000,470.2934,494.7943,471.2486,510.5563,493.8996,493.7573,591.2117,592.6056,580.6567,555.4465,585.9110,575.5860,554.5307,623.7597,606.5698


14. Initial Risk Flags

In [15]:
print("\n🚨 INITIAL RISK FLAGS")

# Potential leakage
suspect_cols = [
    "Lender_portion_to_be_repaid"
]

for col in suspect_cols:
    if col in train.columns:
        print(f"⚠️ Potential leakage column: {col}")

# Check if repayment > amount (interest)
print("\nChecking repayment vs amount:")
check = (train["Total_Amount_to_Repay"] < train["Total_Amount"]).sum()
print("Repay < amount cases:", check)


🚨 INITIAL RISK FLAGS
⚠️ Potential leakage column: Lender_portion_to_be_repaid

Checking repayment vs amount:
Repay < amount cases: 63


In [16]:
# ======================
# Data Type Corrections
# ======================

def fix_dtypes(df):
    df = df.copy()
    
    # Convert dates
    df["disbursement_date"] = pd.to_datetime(df["disbursement_date"], errors="coerce")
    df["due_date"] = pd.to_datetime(df["due_date"], errors="coerce")
    
    # Convert categoricals
    categorical_cols = ["country_id", "loan_type", "New_versus_Repeat"]
    
    for col in categorical_cols:
        df[col] = df[col].astype("category")
    
    return df


train = fix_dtypes(train)
test = fix_dtypes(test)

print("✅ Data types corrected")

✅ Data types corrected


In [17]:
train.dtypes

ID                                        str
customer_id                             int64
country_id                           category
tbl_loan_id                             int64
lender_id                               int64
loan_type                            category
Total_Amount                          float64
Total_Amount_to_Repay                 float64
disbursement_date              datetime64[us]
due_date                       datetime64[us]
duration                                int64
New_versus_Repeat                    category
Amount_Funded_By_Lender               float64
Lender_portion_Funded                 float64
Lender_portion_to_be_repaid           float64
target                                  int64
dtype: object

**Target by group analysis**

In [18]:
print("\n📊 Default rate by loan_type")
display(train.groupby("loan_type")["target"].mean().sort_values(ascending=False))

print("\n📊 Default rate by New vs Repeat")
display(train.groupby("New_versus_Repeat")["target"].mean())


📊 Default rate by loan_type


loan_type
Type_20   0.8333
Type_12   0.5000
Type_17   0.4545
Type_2    0.3889
Type_21   0.1667
Type_6    0.1222
Type_4    0.1172
Type_7    0.0818
Type_5    0.0786
Type_10   0.0667
Type_9    0.0274
Type_1    0.0089
Type_11   0.0000
Type_19   0.0000
Type_18   0.0000
Type_13   0.0000
Type_14   0.0000
Type_16   0.0000
Type_15   0.0000
Name: target, dtype: float64


📊 Default rate by New vs Repeat


New_versus_Repeat
New Loan      0.1993
Repeat Loan   0.0169
Name: target, dtype: float64

**Customer risk signal**

In [19]:
customer_default = train.groupby("customer_id")["target"].agg(["mean", "count"])
display(customer_default.describe())

,mean,count
count,6517.0000,6517.0000
mean,0.0855,10.5346
std,0.2482,12.9728
min,0.0000,1.0000
25%,0.0000,2.0000
50%,0.0000,6.0000
75%,0.0000,14.0000
max,1.0000,206.0000


**Time Distribution**

In [20]:
train["year"] = train["disbursement_date"].dt.year

print("\nLoans per year:")
display(train["year"].value_counts())


Loans per year:


year
2022    64491
2024     2929
2023     1197
2021       37
Name: count, dtype: int64

**Outlier detection**

In [21]:
for col in ["Total_Amount", "Total_Amount_to_Repay"]:
    print(f"\n{col} percentiles:")
    print(train[col].quantile([0.5, 0.9, 0.95, 0.99, 0.999]))


Total_Amount percentiles:
0.5000     5248.0000
0.9000    25000.0000
0.9500    42837.4000
0.9900   114884.5800
0.9990   790609.0000
Name: Total_Amount, dtype: float64

Total_Amount_to_Repay percentiles:
0.5000     5299.0000
0.9000    25746.2000
0.9500    44231.4000
0.9900   119489.9900
0.9990   835303.7776
Name: Total_Amount_to_Repay, dtype: float64


**Correlation with target**

In [22]:
corr = train.corr(numeric_only=True)["target"].sort_values(ascending=False)
display(corr)

target                         1.0000
duration                       0.1721
year                           0.1254
Lender_portion_to_be_repaid    0.1107
Amount_Funded_By_Lender        0.1038
Total_Amount                   0.0752
Total_Amount_to_Repay          0.0739
Lender_portion_Funded          0.0458
tbl_loan_id                    0.0339
customer_id                    0.0282
lender_id                     -0.1581
Name: target, dtype: float64

**Check data leakage stastistally**

In [23]:
print("\nLeakage test:")
display(train.groupby("target")["Lender_portion_to_be_repaid"].mean())


Leakage test:


target
0    2408.5783
1   13072.3100
Name: Lender_portion_to_be_repaid, dtype: float64

**Train Vs Test Distribution**

In [24]:
for col in ["Total_Amount", "duration"]:
    print(f"\n{col} comparison:")
    print("Train mean:", train[col].mean())
    print("Test mean:", test[col].mean())


Total_Amount comparison:
Train mean: 14592.70483322166
Test mean: 15366.443869527804

duration comparison:
Train mean: 8.503568619454073
Test mean: 13.682209314832741


**Economic Data Usability check**

In [25]:
print("\nEconomic missing per country:")
display(econ.groupby("Country").apply(lambda x: x.isnull().mean().mean()))


Economic missing per country:


Country
Cote d'Ivoire   0.2361
Ghana           0.3704
Kenya           0.0509
dtype: float64

**LOAN TYPE SUPPORT SIZE** 

Helps detect Rare categories causing fake high default rates

In [26]:
loan_stats = train.groupby("loan_type")["target"].agg(["mean", "count"])
display(loan_stats.sort_values("mean", ascending=False))

,mean,count
loan_type,,
Type_20,0.8333,18
Type_12,0.5000,4
Type_17,0.4545,11
Type_2,0.3889,360
Type_21,0.1667,6
Type_6,0.1222,360
Type_4,0.1172,1203
Type_7,0.0818,2749
Type_5,0.0786,1514


🚨 CRITICAL INSIGHT

❗ High default rates = LOW SAMPLE SIZE

+ Type_20 → only 18 samples

+ Type_12 → only 4 samples

👉 These are statistical illusions, not real signals


If used directly:

+ Model will overfit rare categories
+ Poor generalization to Ghana

FINAL DECISION -> Group rare loan types

In [33]:
rare_threshold = 100

loan_counts = train["loan_type"].value_counts()
rare_types = loan_counts[loan_counts < rare_threshold].index

# Add "Other" to categories first
train["loan_type"] = train["loan_type"].cat.add_categories(["Other"])
test["loan_type"] = test["loan_type"].cat.add_categories(["Other"])

# Now safely replace
train["loan_type_grouped"] = train["loan_type"].replace(rare_types, "Other")
test["loan_type_grouped"] = test["loan_type"].replace(rare_types, "Other")

**CUSTOMER HISTORY CHECK (LEAKAGE RISK)**

In [27]:
train_sorted = train.sort_values("disbursement_date")

train_sorted["customer_past_defaults"] = (
    train_sorted.groupby("customer_id")["target"]
    .cumsum()
    .shift()
)

display(train_sorted["customer_past_defaults"].describe())

count   68653.0000
mean        0.0277
std         0.2577
min         0.0000
25%         0.0000
50%         0.0000
75%         0.0000
max        15.0000
Name: customer_past_defaults, dtype: float64

**DEFAULT RATE OVER TIME**

In [28]:
display(train.groupby("year")["target"].mean())

year
2021   0.4865
2022   0.0133
2023   0.1136
2024   0.0850
Name: target, dtype: float64

**LOAN SIZE VS DEFAULT**

In [29]:
train["amount_bin"] = pd.qcut(train["Total_Amount"], q=5)

display(train.groupby("amount_bin")["target"].mean())

amount_bin
(4.999, 1938.6]         0.0156
(1938.6, 4140.0]        0.0093
(4140.0, 6949.8]        0.0170
(6949.8, 14120.8]       0.0178
(14120.8, 20000000.0]   0.0318
Name: target, dtype: float64

**DURATION VS DEFAULT**

In [30]:
display(train.groupby("duration")["target"].mean().head(10))

duration
1    0.0000
3    0.0000
4    0.0000
5    0.0000
7    0.0126
8    0.0000
12   0.0000
14   0.0820
15   0.0000
16   0.0000
Name: target, dtype: float64

# 📊 Data Validation & Sanity Check — Summary Report

## 1. 🎯 Objective

The purpose of this notebook was to conduct a comprehensive data validation and sanity check of all datasets provided for the loan default prediction task. This step ensures:

* Data integrity and consistency
* Identification of potential data leakage
* Understanding of dataset structure and relationships
* Early detection of modeling risks
* Establishment of a reliable foundation for feature engineering and modeling

---

## 2. 📁 Datasets Overview

The project utilizes three primary datasets:

* **Train Dataset**: Contains labeled loan records with default outcomes
* **Test Dataset**: Contains unlabeled loan records for prediction
* **Economic Indicators Dataset**: Macroeconomic variables across countries and years

---

## 3. ✅ Key Validation Findings

### 3.1 Missing Values

* **Train Dataset**: No missing values
* **Test Dataset**: No missing values
* **Economic Dataset**:

  * Significant missingness (up to ~48%)
  * Higher missing rates in recent years and certain countries (e.g., Ghana)

📌 **Implication**:

* Core datasets are clean and ready for modeling
* Economic data requires careful handling (imputation or missing indicators)

---

### 3.2 Duplicate Checks

* No duplicate rows detected
* No duplicate IDs detected

📌 **Implication**:

* Dataset integrity is strong
* No need for deduplication

---

### 3.3 Data Types

* Numerical and categorical columns are correctly identified
* Date columns were initially stored as strings and converted to datetime
* Categorical columns require explicit handling (`category` dtype)

📌 **Implication**:

* Data types are now aligned for efficient processing and modeling

---

## 4. 🧠 Structural Insights

### 4.1 Entity Relationships

* **Customers**: ~6,500
* **Loans**: ~66,000
* **Lenders**: 4

#### Key Observations:

* Customers have multiple loans (mean ≈ 10 loans per customer)
* Loans are typically funded by a single lender (multi-lender cases are rare)

📌 **Implication**:

* Strong **customer-level behavioral signal**
* Limited importance of lender aggregation

---

## 5. ⚠️ Target Variable Analysis

* Default rate: **~1.83%**
* Highly imbalanced dataset

📌 **Implication**:

* Accuracy is not a reliable metric
* Requires:

  * Stratified validation
  * Class weighting
  * Threshold optimization

---

## 6. 🚨 Data Leakage Detection

### Identified Risk:

* `Lender_portion_to_be_repaid`

### Evidence:

* Strong statistical separation between default and non-default classes

📌 **Conclusion**:

* This feature likely encodes post-outcome information
* **Will be removed from modeling**

---

## 7. 📊 Feature-Level Insights

### 7.1 Loan Type

* Some loan types show extremely high default rates
* These categories have very low sample sizes

📌 **Implication**:

* High variance / unreliable signals
* Rare categories must be grouped

---

### 7.2 New vs Repeat Loans

* New loans: ~19.9% default rate
* Repeat loans: ~1.69% default rate

📌 **Implication**:

* One of the strongest predictive features
* Reflects borrower trust and history

---

### 7.3 Customer Behavior

* Most customers never default
* Some customers consistently default

📌 **Implication**:

* Strong opportunity for **customer-level feature engineering**
* Must ensure time-aware computation to avoid leakage

---

### 7.4 Loan Amount

* Larger loans show higher default rates
* Distribution is highly skewed with extreme outliers

📌 **Implication**:

* Consider log transformations or tree-based models

---

### 7.5 Loan Duration

* Longer durations are associated with higher default rates

📌 **Implication**:

* Key risk indicator
* Strong predictive feature

---

### 7.6 Temporal Patterns

* Data heavily concentrated in 2022
* Smaller sample sizes in other years
* Some years (e.g., 2021) show inflated default rates due to low sample size

📌 **Implication**:

* Time-based bias exists
* Temporal features must be handled carefully

---

## 8. 🌍 Distribution Shift (Train vs Test)

### Observed Differences:

* **Duration**:

  * Train mean ≈ 8.5 days
  * Test mean ≈ 13.7 days

📌 **Implication**:

* Significant distribution shift
* Model must generalize beyond training distribution

---

## 9. 🌍 Economic Data Assessment

* Contains 9 macroeconomic indicators
* Varying levels of missingness by country:

  * Kenya: low missingness (~5%)
  * Ghana: high missingness (~37%)

📌 **Implication**:

* Useful for domain generalization
* Requires:

  * Careful merging
  * Missing value handling strategy

---

## 10. ⚠️ Key Risks Identified

1. **Severe Class Imbalance**
2. **Potential Data Leakage**
3. **Distribution Shift (Train vs Test)**
4. **Sparse and Noisy Categories**
5. **Temporal Bias**
6. **Economic Data Missingness**

---

## 11. ✅ Strategic Decisions

Based on validation findings, the following decisions are made:

* ❌ Drop leakage-prone features:

  * `Lender_portion_to_be_repaid`

* 🔧 Group rare categorical values:

  * `loan_type`

* 🧠 Prioritize customer-level features:

  * Behavioral and historical patterns

* ⚖️ Handle class imbalance:

  * Use F1-optimized strategies

* 🔄 Use grouped validation:

  * `GroupKFold` with `customer_id`

* 🌍 Incorporate economic indicators cautiously

---

## 12. 🚀 Next Steps

The next phase focuses on **Feature Engineering**, where insights from this validation stage will be translated into:

* Customer-level features
* Loan-level transformations
* Temporal features
* Economic feature integration

A reusable, production-grade feature engineering pipeline will be implemented to ensure:

* Consistency between training and inference
* Prevention of data leakage
* Scalability for experimentation

---

## 🧠 Final Remark

This validation step revealed that the problem is not a simple tabular classification task, but rather a **behavioral credit risk modeling problem** involving:

* Multi-level data relationships
* Temporal dynamics
* Domain generalization challenges

A thoughtful and structured approach will be critical to building a robust and generalizable model.
